packages/core-ethereum/crates/core-ethereum-misc/src/chain.rs

```diff
@@ -1,12 +1,112 @@
     Result,
 };
-use core_crypto::types::Hash;
+use bindings::{
+    hopr_announcements::BindKeysAnnounceCall,
+    hopr_channels::{
+        CloseIncomingChannelCall, CompactSignature, FundChannelCall, InitiateOutgoingChannelClosureCall,
+        RedeemTicketCall, RedeemableTicket, TicketData, Vrfparameters,
+    },
+};
+use core_crypto::{
+    derivation::derive_vrf_parameters,
+    keypairs::{ChainKeypair, Keypair, OffchainKeypair},
+    types::{Hash, VrfParameters},
+};
 use core_ethereum_db::traits::HoprCoreEthereumDbActions;
-use core_types::acknowledgement::AcknowledgedTicket;
+use core_types::{account::AccountSignature, acknowledgement::AcknowledgedTicket, channels::generate_channel_id};
+use ethers::types::{Address as EthereumAddress, H256, U256};
+use multiaddr::Multiaddr;
 use utils_log::debug;
-use utils_types::primitives::Address;
+use utils_types::{
+    primitives::{Address, Balance, BalanceType},
+    traits::BinarySerializable,
+};
+pub fn announce(
+    offchain_keypair: &OffchainKeypair,
+    chain_key: &Address,
+    announced_multiaddr: &Multiaddr,
+) -> BindKeysAnnounceCall {
+    let account_sig = AccountSignature::new(offchain_keypair, chain_key);
+
+    let serialized_signature = account_sig.signature.to_bytes();
+
+    BindKeysAnnounceCall {
+        base_multiaddr: announced_multiaddr.to_string(),
+        ed_25519_pub_key: H256::from_slice(&offchain_keypair.public().to_bytes()).into(),
+        ed_25519_sig_0: H256::from_slice(&serialized_signature[0..32]).into(),
+        ed_25519_sig_1: H256::from_slice(&serialized_signature[32..64]).into(),
+    }
+}
+
+pub fn fund_channel(dest: &Address, amount: Balance) -> Result<FundChannelCall> {
+    if amount.balance_type() != BalanceType::HOPR {
+        return Err(InvalidArguments(
+            "Invalid balance type. Expected a HOPR balance.".into(),
+        ));
+    }
+    Ok(FundChannelCall {
+        amount: amount.value().as_u128(),
+        account: EthereumAddress::from_slice(&dest.to_bytes()),
+    })
+}
+
+pub fn close_incoming_channel(source: &Address) -> CloseIncomingChannelCall {
+    CloseIncomingChannelCall {
+        source: EthereumAddress::from_slice(&source.to_bytes()),
+    }
+}
+
+pub fn initiate_outgoing_channel_closure(destination: &Address) -> InitiateOutgoingChannelClosureCall {
+    InitiateOutgoingChannelClosureCall {
+        destination: EthereumAddress::from_slice(&destination.to_bytes()),
+    }
+}
+
+pub fn redeem_ticket(chain_keypair: &ChainKeypair, acked_ticket: &AcknowledgedTicket) -> Result<RedeemTicketCall> {
+    let channel_id = generate_channel_id(&acked_ticket.signer, &chain_keypair.public().to_address());
+
+    let serialized_signature = match acked_ticket.ticket.signature {
+        Some(ref signature) => signature.to_bytes(),
+        None => return Err(InvalidArguments("Acknowledged ticket must be signed".into())),
+    };
+
+    // BIG TODO
```

Comment:
Still TODO?

Reply:
in the making, needs many structs to be slightly changed

Reply:
well, it's a small TODO now